### PR TITLE
[SPARK-24795][Core][FOLLOWUP] Kill all running tasks when a task in a barrier stage fail

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

Kill all running tasks when a task in a barrier stage fail in the middle. `TaskScheduler`.`cancelTasks()` will also fail the job, so we implemented a new method `killAllTaskAttempts()` to just kill all running tasks of a stage without cancel the stage/job.

## How was this patch tested?

Add new test cases in `TaskSchedulerImplSuite`.

### CODE REVIEW

**Comment by holdensmagicalunicorn** [comment]
@jiangxb1987, thanks! I am a bot who has found some folks who might be able to help with the review:@gatorsmile, @mateiz and @kayousterhout

File: core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala [hunk 1]
```diff
@@ -51,7 +51,7 @@ private[spark] trait TaskScheduler {
   // Submit a sequence of tasks to run.
   def submitTasks(taskSet: TaskSet): Unit
 
-  // Cancel a stage.
+  // Kill all the tasks in a stage and fail the stage and all the jobs that depend on the stage.
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskScheduler.scala]: is it guaranteed to work for any backend like YARN, Mesos, K8s?

**Comment by mengxr** [comment]
LGTM

**Comment by jiangxb1987** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**